### ISSUE TITLE
eprof causes segmentation fault upon garbage_collect on OTP 22

### ISSUE THREAD

**Issue by dszoboszlay** [erlang/otp#4635]
I found a segfault on OTP 22 when running `eprof`.

* OTP: 22.3.4.16 (I haven't tested later versions yet)
* OS: Ubuntu 20.04.2 LTS
* Arch: x86_64

Take the below sample program:

```erlang
-module(segfault).
-export([ start/0
        , worker/0
        ]).

start() ->
  Pid = spawn(?MODULE, worker, []),
  eprof:start(),
  eprof:start_profiling([Pid]),
  timer:sleep(1000),
  eprof:stop_profiling(),
  eprof:stop().

worker() ->
  receive
    quit -> ok
  after 100 ->
    erlang:garbage_collect(),
    ?MODULE:?FUNCTION_NAME()
  end.
```

Running the program causes a segmentation fault:

```shell
$ erlc segfault.erl && erl -s segfault
Erlang/OTP 22 [erts-10.7.2.8] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1]

Eshell V10.7.2.8  (abort with ^G)
1> Segmentation fault (core dumped)
```

This is what I see in `gdb`:

```
Thread 8 "4_scheduler" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ff403376700 (LWP 360149)]
0x000055a254d687fb in fixup_cp_before_trace (return_to_trace=<synthetic pointer>, c_p=0x7ff405a02c98) at beam/beam_bp.c:670
670	            BeamInstr w = *cp_val(*cpp);
(gdb) print *c_p
$1 = {common = {id = 2817498547491, refc = {atmc = {counter = 1}, sint = 1}, tracer = 140686055703121, trace_flags = 1088, timer = {counter = 0}, u = {alive = {
        started_interval = 0, reg = 0x0, links = 0x0, lt_monitors = 0x0, monitors = 0x0}, release = {later = 0, func = 0x0, data = 0x0, next = 0x0}}}, htop = 0x7ff401815840, 
  stop = 0x7ff401815f70, heap = 0x7ff401815840, hend = 0x7ff401815f88, abandoned_heap = 0x0, heap_sz = 233, min_heap_size = 233, min_vheap_size = 46422, max_heap_size = 3, 
  arity = 0, arg_reg = 0x7ff405a02d48, max_arg_reg = 6, def_arg_reg = {140685976288104, 208907, 18446744073709551611, 0, 0, 4000}, cp = 0x55a255054708 <beam_return_time_trace>, 
  i = 0x7ff4040ff8d8, catches = 0, fcalls = 3997, rcount = 0, schedule_count = 0, reds = 5, flags = 0, group_leader = 2233382994963, fvalue = 18446744073709551611, freason = 0, 
  ftrace = 18446744073709551611, next = 0x0, sig_qs = {first = 0x0, last = 0x7ff405a02dd8, save = 0x7ff405a02dd8, cont = 0x0, cont_last = 0x7ff405a02df0, nmsigs = {next = 0x0, 
      last = 0x0}, saved_last = 0x0, len = 0, flags = 2}, bif_timers = 0x0, dictionary = 0x0, seq_trace_clock = 0, seq_trace_lastcnt = 0, seq_trace_token = 18446744073709551611, 
  u = {real_proc = 0x89e0b, terminate = 0x89e0b, initial = {module = 564747, function = 208907, arity = 0}}, current = 0x7ff4040ff8c0, parent = 2714419332339, static_flags = 0, 
  high_water = 0x7ff401815840, old_hend = 0x0, old_htop = 0x0, old_heap = 0x0, gen_gcs = 0, max_gen_gcs = 65535, off_heap = {first = 0x0, overhead = 0}, mbuf = 0x0, 
  live_hf_end = 0xfffffffffffffff8, msg_frag = 0x0, mbuf_sz = 0, psd = {counter = 140685981587736}, bin_vheap_sz = 46422, bin_old_vheap_sz = 46422, bin_old_vheap = 0, 
  sys_task_qs = 0x0, dirty_sys_tasks = 0x0, state = {counter = 41002}, dirty_state = {counter = 0}, sig_inq = {first = 0x0, last = 0x7ff405a02f10, len = 0, nmsigs = {next = 0x0, 
      last = 0x0}}, trace_msg_q = 0x0, lock = {flags = {counter = 1}, queue = {0x0, 0x0, 0x0, 0x0, 0x0}}, scheduler_data = 0x7ff404463940, run_queue = {
    counter = 140686020325504}}
(gdb) print *c_p->cp
$2 = 94155696147326
(gdb) print c_p->cp
$3 = (BeamInstr *) 0x55a255054708 <beam_return_time_trace>
(gdb) print cpp
$4 = (Eterm *) 0x7ff401815f70
(gdb) print &E[0]
$5 = (Eterm *) 0x7ff401815f70
(gdb) print (BeamInstr*)(E[0])
$6 = (BeamInstr *) 0xfffffffffffffffb
(gdb) print (BeamInstr*)(E[1])
$7 = (BeamInstr *) 0x7ff401a43b80
(gdb) print (BeamInstr*)(E[2])
$8 = (BeamInstr *) 0x55a255054748 <beam_apply+8>
```

So my understanding is that the crash happens [here](https://github.com/erlang/otp/blob/OTP-22.3.4.16/erts/emulator/beam/beam_bp.c#L670).

* The function should have entered the `BeamIsOpCode(w, op_i_return_time_trace)` branch (L663).
* The crash happens in the first iteration within the `for (;;)` loop.
* I don't  fully understand what the code intends to do, but  after checking the `cp` is one of the ops it's looking for it seems to expect that the same op is on the top of the stack. But when I print the stack with `gdb` the top of the stack is `-5` instead. The first op on the stack comes two words later, and it's not the same as the `cp`.
* The `erlang:garbage_collect()` call seems to be important, if I remove it, the segfault goes away.

**Comment by garazdawi** [comment]
I was able to reproduce this segfault in the OTP-22 docker image, but unable to do so in both OTP-23 and OTP-24-rc1. So this bug seems to have been fixed in later releases.

**Comment by dszoboszlay** [comment]
I'm not surprised, `fixup_cp_before_trace` changed a lot between OTP-22 and OTP-23.

**Comment by mikpe** [comment]
I tried bisecting this, assuming we were missing a fix from the master branch, but git bisect came up empty. It appears the bug only exists on the OTP-22 branch, and was introduced by:

`# first bad commit: [1f5fb6de3e20cf433606c8d8c33a9813d6337aa2] erts: Fix call_time trace return`

This also explains the `-5` you found, since that's `NIL` and the commit changed the code to store `NIL` instead of a continuation pointer to the stack in some cases.

I'll defer to the OTP team for further investigation.

**Comment by sverker** [comment]
Fixed in OTP-22.3.4.17.
